### DISCUSSION TITLE
Shapecast can't adjust mCenterOfMassStart

### DISCUSSION THREAD

**Discussion by DennisSmolek** [jrouwe/JoltPhysics.js#156] [Q&A] [answered]
I'm working on my shapecaster and having a few odd issues.

First, the values are all read-only:
```ts
  class ShapeCast {
        constructor(inShape: Shape, inScale: Vec3, inCenterOfMassStart: Mat44, inDirection: Vec3);
        get_mShape(): Shape;
        set_mShape(mShape: Shape): void;
        readonly mShape: Shape;
        get_mScale(): Vec3;
        set_mScale(mScale: Vec3): void;
        readonly mScale: Vec3;
        get_mCenterOfMassStart(): Mat44;
        set_mCenterOfMassStart(mCenterOfMassStart: Mat44): void;
        readonly mCenterOfMassStart: Mat44;
        get_mDirection(): Vec3;
        set_mDirection(mDirection: Vec3): void;
        readonly mDirection: Vec3;
        GetPointOnRay(inFraction: number): Vec3;
    }
```
- This throws typescript errors when trying to adjust these values. 

Raycast doesn't have this:
```ts
    class RayCast {
        constructor();
        constructor(inOrigin: Vec3, inDirection: Vec3);
        Transformed(inTransform: Mat44): RayCast;
        Translated(inTranslation: Vec3): RayCast;
        GetPointOnRay(inFraction: number): Vec3;
        get_mOrigin(): Vec3;
        set_mOrigin(mOrigin: Vec3): void;
        mOrigin: Vec3;
        get_mDirection(): Vec3;
        set_mDirection(mDirection: Vec3): void;
        mDirection: Vec3;
    }
```
- So I tried to use `set_mCenterOfMassStart(MyMat44)` which fails at runtime saying there isn't a function by that name. 

- So then I tried to ignore typescript and just set `mCenterOfMassStart = MyMat44;` which failed, saying there's a getter but no setter..

- So I decided to try adjusting the existing `Mat44` with `shapecast.mCenterOfMassStart.SetTranslation(myOriginVec3);`
This works! and when I do `shapecast.mCenterOfMassStart.GetTranslation()` it's the correct value.

Unfortunately, the shapecast ignores any changed translation and only fires from the initial position used in the constructor. 

What I've done in the meantime is discard the shapecast after each use and generate a new one each time.

Is a shapecast a single use function? Raycast can be done over and over but maybe I missed something.

**Comment by jrouwe** [comment, marked as answer]
The shape cast internally caches a bounding box that is used during casting, that's why you cannot change its properties after construction.

You can do multiple casts with the same ShapeCast but the object itself is indeed read only.
